packages/flutter_tools/lib/src/base/os.dart

```diff
@@ -538,10 +538,26 @@ class _WindowsUtils extends OperatingSystemUtils {
         continue;
       }
 
-      final File destFile = _fileSystem.file(_fileSystem.path.join(
-        targetDirectory.path,
-        archiveFile.name,
-      ));
+      final File destFile = _fileSystem.file(
+        _fileSystem.path.normalize(
+          _fileSystem.path.join(
+            targetDirectory.path,
+            archiveFile.name,
+          ),
+        ),
+      );
+
+      // Validate that the destFile is within the targetDirectory we want to
+      // extract to.
+      //
+      // See https://snyk.io/research/zip-slip-vulnerability for more context.
+      if (!destFile.path.startsWith(targetDirectory.path)) {
```

Comment:
I wish there a better way to check file system entity ancestry, but I don't see a helper to do that anywhere.  Should probably use `absolute`?
```suggestion
      if (!destFile.absolute.path.startsWith(targetDirectory.absolute.path)) {
```

Reply:
> I wish there a better way to check file system entity ancestry

Note the `/windows-x64` and `\windows-x64` path separator difference in the test failures:
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8825193847068836465/+/u/run_build_aar_module_test/test_stdout
```
[build_aar_module_test] [STDOUT] stderr: Bad state: Tried to extract the file C:\b\s\w\ir\x\w\recipe_cleanup\tmp625z6idb\flutter sdk\bin\cache\artifacts\engine\android-arm-profile\windows-x64\gen_snapshot.exe outside of the target directory C:\b\s\w\ir\x\w\recipe_cleanup\tmp625z6idb\flutter sdk\bin\cache\artifacts\engine\android-arm-profile/windows-x64
```

Reply:
+1 to adding `absolute`. Consider if `targetDirectory` were `.`.

Reply:
Instead used `path.canonicalize`, as it handles not only absolute paths, but also casing on Windows (which I forgot about):

https://pub.dev/documentation/path/latest/path/Context/canonicalize.html

> This is guaranteed to return the same path for two different input paths if and only if both input paths point to the same location. Unlike normalize, it returns absolute paths when possible and canonicalizes ASCII case on Windows.